cpp/src/ice2slice/Gen.cpp

```diff
@@ -43,13 +71,30 @@ namespace
 
         if (auto metadata = p->getMetadataArgs("cs:namespace"))
         {
-            hasCSharpNamespaceAttribute = true;
-            return *metadata + "." + csharpNamespace;
+            return *metadata;
         }
         else
         {
-            hasCSharpNamespaceAttribute = false;
-            return csharpNamespace;
+            return "";
+        }
+    }
+
+    static string getCSharpNamespace(const ContainedPtr& cont, bool& hasModuleRemapping)
+    {
+        // Module remapping is required if the module has a cs:namespace attribute or any of the modules has
+        // a cs:identifier
+        string csNamespacePrefix = getCSharpNamespacePrefix(cont);
+        hasModuleRemapping = !csNamespacePrefix.empty() || cont->scope() != getCSharpMappedScope(cont, "::");
```

Comment:
`|| cont->scope() != getCSharpMappedScope(cont, "::");`

I find the logic too round-about. We remap the module when the Slice scope of cont is different from the mapped scope of cont using `::` separators?

What about a simpler comparison: take the default mapping and compare it with the actual mapping?
And to do even simpler, this comparison could be performed by the unique caller: we just need one function that returns the actual mapping and another that returns the default mapping. Then no hasModuleRemapping, no separator parameter ... so much simpler.